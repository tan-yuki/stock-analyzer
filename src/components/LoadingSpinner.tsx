import React from 'react';

export const LoadingSpinner: React.FC = () => {
  return (
    <div className="loading">
      <div className="spinner"></div>
      <p>データを取得中...</p>
    </div>
  );
};