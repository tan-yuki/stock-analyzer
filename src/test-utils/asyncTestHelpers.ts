import { vi, expect } from 'vitest'

// 順序を保証する非同期テストヘルパー
export const sequentialAsync = async <T>(
  operations: (() => Promise<T>)[]
): Promise<T[]> => {
  const results: T[] = []
  
  for (const operation of operations) {
    const result = await operation()
    results.push(result)
  }
  
  return results
}

// モック呼び出し順序の検証ヘルパー
export const verifyCallOrder = (mockFn: ReturnType<typeof vi.fn>, expectedCalls: unknown[][]) => {
  const actualCalls = mockFn.mock.calls
  
  expect(actualCalls).toHaveLength(expectedCalls.length)
  
  expectedCalls.forEach((expectedCall, index) => {
    expect(actualCalls[index]).toEqual(expectedCall)
  })
}

// タイムアウト付きPromise実行
export const withTimeout = <T>(
  promise: Promise<T>,
  timeoutMs: number,
  errorMessage?: string
): Promise<T> => {
  return Promise.race([
    promise,
    new Promise<never>((_, reject) => 
      setTimeout(() => reject(new Error(errorMessage || `Operation timed out after ${timeoutMs}ms`)), timeoutMs)
    )
  ])
}

// リトライ機能付きテスト実行
export const retryOperation = async <T>(
  operation: () => Promise<T>,
  maxRetries: number = 3,
  delay: number = 100
): Promise<T> => {
  let lastError: Error
  
  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      return await operation()
    } catch (error) {
      lastError = error as Error
      if (attempt < maxRetries) {
        await new Promise(resolve => setTimeout(resolve, delay))
      }
    }
  }
  
  throw lastError!
}