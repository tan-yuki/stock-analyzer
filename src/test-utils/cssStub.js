// CSS stub for testing
module.exports = {}