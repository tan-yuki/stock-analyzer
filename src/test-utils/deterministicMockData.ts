// 確定的なテストデータ生成（Flakyテスト対策）
export const createDeterministicStockData = (symbol: string, period: string) => {
  // シードベースの確定的な価格生成
  const seed = symbol.charCodeAt(0) + period.length
  let randomSeed = seed
  
  const seededRandom = () => {
    randomSeed = (randomSeed * 9301 + 49297) % 233280
    return randomSeed / 233280
  }
  
  const basePrice = 150 // 固定ベース価格
  const dates = generateTradingDates(period)
  
  return dates.map((date, _index) => {
    const variance = (seededRandom() - 0.5) * 10 // ±5の変動
    return {
      date,
      price: parseFloat((basePrice + variance).toFixed(2))
    }
  })
}

const generateTradingDates = (period: string): Date[] => {
  const now = new Date('2025-06-01') // 固定日付でタイムゾーン問題を回避
  const dates: Date[] = []
  const periodDays = { '1mo': 30, '3mo': 90, '6mo': 180, '1y': 365, '2y': 730 }
  const days = periodDays[period as keyof typeof periodDays] || 30
  
  for (let i = days; i >= 0; i--) {
    const date = new Date(now)
    date.setDate(date.getDate() - i)
    
    // 平日のみ追加（土日スキップで確定的な取引日生成）
    if (date.getDay() !== 0 && date.getDay() !== 6) {
      dates.push(date)
    }
  }
  
  return dates
}

// 特定銘柄の確定的なベース価格を設定
export const getDeterministicBasePrice = (symbol: string): number => {
  const basePrices: Record<string, number> = {
    'AAPL': 150.00,
    'GOOGL': 2800.00,
    'TSLA': 800.00,
    'MSFT': 420.00,
    'AMZN': 3400.00,
    'NVDA': 900.00,
    'META': 500.00
  }
  
  return basePrices[symbol] || 150.00
}