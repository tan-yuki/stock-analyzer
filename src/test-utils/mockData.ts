import { StockPrice, StockData, StockAnalysis } from '../types'

export const mockStockPrices: StockPrice[] = [
  { date: new Date('2024-01-01'), price: 150.00 },
  { date: new Date('2024-01-02'), price: 152.50 },
  { date: new Date('2024-01-03'), price: 148.75 },
  { date: new Date('2024-01-04'), price: 151.25 },
  { date: new Date('2024-01-05'), price: 155.00 },
  { date: new Date('2024-01-08'), price: 153.75 },
  { date: new Date('2024-01-09'), price: 157.25 },
  { date: new Date('2024-01-10'), price: 159.50 },
]

export const mockStockData: StockData = {
  symbol: 'AAPL',
  companyName: 'Apple Inc.',
  prices: mockStockPrices,
  currentPrice: 159.50,
  previousPrice: 157.25,
}

export const mockStockAnalysis: StockAnalysis = {
  max: 159.50,
  min: 148.75,
  average: 153.46875,
  volatility: 12.34,
  totalReturn: 6.33,
}

export const mockAlphaVantageResponse = {
  'Meta Data': {
    '1. Information': 'Daily Prices (open, high, low, close) and Volumes',
    '2. Symbol': 'AAPL',
    '3. Last Refreshed': '2024-01-10',
    '4. Output Size': 'Compact',
    '5. Time Zone': 'US/Eastern',
  },
  'Time Series (Daily)': {
    '2024-01-10': {
      '1. open': '158.0000',
      '2. high': '160.0000',
      '3. low': '157.0000',
      '4. close': '159.5000',
      '5. volume': '50000000',
    },
    '2024-01-09': {
      '1. open': '156.0000',
      '2. high': '158.0000',
      '3. low': '155.0000',
      '4. close': '157.2500',
      '5. volume': '45000000',
    },
    '2024-01-08': {
      '1. open': '154.0000',
      '2. high': '156.0000',
      '3. low': '153.0000',
      '4. close': '153.7500',
      '5. volume': '42000000',
    },
  },
}

export const mockCompanyOverview = {
  Symbol: 'AAPL',
  Name: 'Apple Inc.',
  Description: 'Apple Inc. designs, manufactures, and markets smartphones...',
  Exchange: 'NASDAQ',
  Currency: 'USD',
}

export const createMockStockPrice = (overrides: Partial<StockPrice> = {}): StockPrice => ({
  date: new Date('2024-01-01'),
  price: 100.00,
  ...overrides,
})

export const createMockStockData = (overrides: Partial<StockData> = {}): StockData => ({
  symbol: 'TEST',
  companyName: 'Test Company Inc.',
  prices: [createMockStockPrice()],
  currentPrice: 100.00,
  previousPrice: 95.00,
  ...overrides,
})