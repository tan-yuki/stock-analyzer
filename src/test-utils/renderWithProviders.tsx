import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'

// カスタムレンダー関数（将来的にContextやProviderが必要になった場合に使用）
function AllTheProviders({ children }: { children: React.ReactNode }) {
  return <>{children}</>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options })

// React Fast Refreshの警告を避けるため、re-exportと関数を分離
export { customRender as render }
export * from '@testing-library/react'