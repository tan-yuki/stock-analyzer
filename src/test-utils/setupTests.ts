import '@testing-library/jest-dom'
import { vi } from 'vitest'

// Chart.js のモック
vi.mock('chart.js', () => {
  const mockChart = vi.fn().mockImplementation(() => ({
    destroy: vi.fn(),
    update: vi.fn(),
    render: vi.fn(),
  }))
  
  mockChart.register = vi.fn()
  
  return {
    Chart: mockChart,
    CategoryScale: vi.fn(),
    LinearScale: vi.fn(),
    PointElement: vi.fn(),
    LineElement: vi.fn(),
    Title: vi.fn(),
    Tooltip: vi.fn(),
    Legend: vi.fn(),
    Filler: vi.fn(),
    register: vi.fn(),
  }
})

vi.mock('react-chartjs-2', () => ({
  Line: vi.fn().mockImplementation(({ data, options }) => {
    const React = require('react')
    return React.createElement('div', {
      'data-testid': 'mock-chart',
      'data-chart-data': JSON.stringify(data),
      'data-chart-options': JSON.stringify(options)
    }, 'Chart Placeholder')
  }),
}))

// Window alert のモック
global.alert = vi.fn()

// Console methods のモック (テスト中のログを抑制)
global.console = {
  ...console,
  warn: vi.fn(),
  error: vi.fn(),
  log: vi.fn(),
}

// Fetch API のモック (MSW使用時は不要だが、フォールバック用)
global.fetch = vi.fn()

// 環境変数のモック
Object.defineProperty(import.meta, 'env', {
  value: {
    VITE_ALPHA_VANTAGE_API_KEY: 'test-api-key'
  },
  writable: true
})

// 確定的な時間設定（Flakyテスト対策）
const FIXED_DATE = new Date('2025-06-01T00:00:00Z')

// Date.nowのモック
const originalDateNow = Date.now
Date.now = vi.fn(() => FIXED_DATE.getTime())

// new Date()のモック（引数なしの場合のみ）
const originalDate = global.Date
global.Date = class extends originalDate {
  constructor(...args: any[]) {
    if (args.length === 0) {
      super(FIXED_DATE.getTime())
    } else {
      super(...args)
    }
  }
  
  static now() {
    return FIXED_DATE.getTime()
  }
} as any