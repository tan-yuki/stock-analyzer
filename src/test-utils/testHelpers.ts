import { vi } from 'vitest'
import { screen, waitFor } from '@testing-library/react'

// MSWの応答を意図的に遅延させるヘルパー
export const createDelayedMSWHandler = (delay: number = 100) => {
  return async () => {
    await new Promise(resolve => setTimeout(resolve, delay))
    // 元のハンドラーロジック
  }
}

// 確定的なタイミングテスト用のヘルパー（Flakyテスト対策）
export const waitForElementToAppear = async (
  getElement: () => HTMLElement | null,
  timeout: number = 5000
) => {
  const startTime = Date.now()
  
  while (Date.now() - startTime < timeout) {
    const element = getElement()
    if (element) return element
    await new Promise(resolve => setTimeout(resolve, 50))
  }
  
  throw new Error(`Element did not appear within ${timeout}ms`)
}

// 状態変化を確実に検出するヘルパー
export const waitForStateChange = async <T>(
  getCurrentState: () => T,
  expectedState: T,
  timeout: number = 5000
) => {
  const startTime = Date.now()
  
  while (Date.now() - startTime < timeout) {
    if (getCurrentState() === expectedState) {
      return true
    }
    await new Promise(resolve => setTimeout(resolve, 50))
  }
  
  throw new Error(`State did not change to expected value within ${timeout}ms`)
}

// より安全な要素待機ヘルパー（Testing LibraryのwaitForを使用）
export const waitForElementSafely = async (
  textOrTestId: string | RegExp,
  options: { timeout?: number; testId?: boolean } = {}
) => {
  const { timeout = 5000, testId = false } = options
  
  return waitFor(() => {
    if (testId) {
      return screen.getByTestId(textOrTestId as string)
    }
    return screen.getByText(textOrTestId)
  }, { timeout })
}

// 複数要素の安全な取得
export const getAllElementsSafely = (
  textOrSelector: string | RegExp,
  options: { testId?: boolean } = {}
) => {
  const { testId = false } = options
  
  if (testId) {
    return screen.getAllByTestId(textOrSelector as string)
  }
  return screen.getAllByText(textOrSelector)
}

// アラート呼び出しの確実な検証
export const verifyAlertCalled = async (
  mockAlert: ReturnType<typeof vi.fn>,
  expectedMessage: string,
  timeout: number = 3000
) => {
  return waitFor(() => {
    expect(mockAlert).toHaveBeenCalledWith(expectedMessage)
  }, { timeout })
}