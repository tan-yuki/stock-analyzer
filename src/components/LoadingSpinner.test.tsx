import { describe, it, expect } from 'vitest'
import { render, screen } from '../test-utils/renderWithProviders'
import { LoadingSpinner } from './LoadingSpinner'

describe('LoadingSpinner', () => {
  it('ローディングメッセージが表示される', () => {
    render(<LoadingSpinner />)

    expect(screen.getByText('データを取得中...')).toBeInTheDocument()
  })

  it('スピナー要素が表示される', () => {
    const { container } = render(<LoadingSpinner />)

    // Check for div elements that should contain the spinner
    const divElements = container.querySelectorAll('div')
    expect(divElements.length).toBeGreaterThanOrEqual(2) // container + spinner
  })

  it('適切な構造で表示される', () => {
    const { container } = render(<LoadingSpinner />)

    // Check that the loading message is displayed
    expect(screen.getByText('データを取得中...')).toBeInTheDocument()
    
    // Check that the component has the expected structure
    const outerDiv = container.firstChild as HTMLElement
    expect(outerDiv).toBeInTheDocument()
    expect(outerDiv.tagName).toBe('DIV')
  })

  it('アクセシビリティ用のロール属性を持つ', () => {
    render(<LoadingSpinner />)

    // ローディング状態をスクリーンリーダーに伝える
    const loadingElement = screen.getByText('データを取得中...')
    expect(loadingElement).toBeInTheDocument()
  })

  it('スナップショットテスト', () => {
    const { container } = render(<LoadingSpinner />)
    expect(container.firstChild).toMatchSnapshot()
  })
})