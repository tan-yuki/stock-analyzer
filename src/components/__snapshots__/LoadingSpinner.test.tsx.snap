// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoadingSpinner > スナップショットテスト 1`] = `
<div
  style="text-align: center; padding: 40px;"
>
  <div
    style="width: 32px; height: 32px; border: 3px solid rgb(240, 240, 240); border-top: 3px solid #2196F3; border-radius: 50%; animation: spin 1s linear infinite; margin: 0px auto 16px;"
  />
  <p>
    データを取得中...
  </p>
</div>
`;
